toolchain/parse/testdata/generics/impl/fail_out_of_line_member.carbon

```diff
@@ -84,13 +74,8 @@ fn C.(Self as Interface).F() {}
 // CHECK:STDOUT:     {kind: 'ClassDefinition', text: '}', subtree_size: 15},
 // CHECK:STDOUT:       {kind: 'FunctionIntroducer', text: 'fn'},
 // CHECK:STDOUT:         {kind: 'IdentifierName', text: 'C'},
-// CHECK:STDOUT:         {kind: 'IdentifierName', text: '(', has_error: yes},
-// CHECK:STDOUT:       {kind: 'QualifiedName', text: '.', subtree_size: 3},
-// CHECK:STDOUT:         {kind: 'TuplePatternStart', text: '('},
-// CHECK:STDOUT:           {kind: 'IdentifierName', text: 'Self', has_error: yes},
-// CHECK:STDOUT:           {kind: 'InvalidParse', text: 'Self', has_error: yes},
-// CHECK:STDOUT:         {kind: 'BindingPattern', text: 'Self', has_error: yes, subtree_size: 3},
-// CHECK:STDOUT:       {kind: 'TuplePattern', text: ')', has_error: yes, subtree_size: 5},
-// CHECK:STDOUT:     {kind: 'FunctionDecl', text: '}', has_error: yes, subtree_size: 10},
+// CHECK:STDOUT:       {kind: 'NameQualifier', text: '.', subtree_size: 2},
+// CHECK:STDOUT:       {kind: 'InvalidParse', text: '(', has_error: yes},
```

Comment:
This InvalidParse seems a little awkward since you document the IdentifierName as bracketing; how does the system distinguish what starts/ends where?

But also, are there other ways to cause this? What does something like `fn *.F() {}` parse like? Does that yield anything that can be identified as the "identifier"?

Reply:
`IdentifierName` is bracketing for a `NameQualifier`, but this isn't part of a `NameQualifier`.

For `fn *.F() {}`, we fail to parse a declaration name at the `*` and stop name parsing there (this is the `???` rule you asked about in `state.def`). This eventually leads to the overall function parsing failing, because `*` isn't a valid start for a return type or a function body, and we skip to the end of the declaration. That results in:
```console
/dev/stdin:1:4: ERROR: `fn` introducer should be followed by a name.
fn *.F() {}
   ^
```
and this parse tree:
```yaml
- filename: /dev/stdin
  parse_tree: [
    {kind: 'FileStart', text: ''},
      {kind: 'FunctionIntroducer', text: 'fn'},
      {kind: 'InvalidParse', text: '*', has_error: yes},
    {kind: 'FunctionDecl', text: '}', has_error: yes, subtree_size: 3},
    {kind: 'FileEnd', text: ''},
  ]
```